### DISCUSSION TITLE
Directus self-hosted + AWS S3 + AWS Cloudfront

### DISCUSSION THREAD

**Discussion by luigigorlero** [directus/directus#17359] [Feature Requests]
Hello!
I'm having troubles getting an optimal configuration of Directus to use S3 + CloudFront + Directus transformations on images. 
As of now, my Directus communicates flawlessly with the S3 bucket, and CloudFront is configured with access to the bucket. 
So, any request to `https://{DIRECTUS_URL}/assets/{ID}` gets the file directly from the S3 bucket (with the possibility of applying transformation, but with the drawbacks of S3 low speeds), while any request to `https://{CLOUDFRONT_URL}/{FILENAME.EXT}` gets the file from the CloudFront CDN. 

What I have not been able to do is configuring the services in such a way that any request to `https://{DIRECTUS_URL}/assets/` gets handled directly by the CDN (if the transformed file is available and in cache). Is that even possible? I was looking into CloudFront custom origins but haven't been able to figure out a solution yet.

Anybody has found themselves here too?

**Comment by chris-lsn** [comment]
I'm facing the same issue of you but I use Cloudflare R2. The storage endpoint doesn't have CDN by default, they are required to setup a custom domain with CDN enabled to access assets instead. Unfortunately, directus doesn't handle this use case, but it is possible to accomplish by modifying the way of getting the stream in the "read" method of '@directus/storage-driver-s3'.

**Reply by Tim-Ganther** [reply]
Unfortunately, directus still doesn't handle this use case. I had a look at the "read" method of '@directus/storage-driver-s3' but I am unsure how I have to modify it to make it work with AWS Cloudfront instead of directly reading from S3. Any tips/hints?

Appreciate any help!

**Reply by martinsotirov** [reply]
@chris-lsn did you figure out how to use transformations on R2 hosted images? On my public R2 domain I only get exposed what's physically created by directus as real files so I can't access dynamic transformations...

**Reply by luigigorlero** [reply]
What you’re mentioning is not possible, since the Transformation Worker lives inside Directus and accessing an R2 public link means completely bypassing Directus. 
Since my first post, I ended up leaving AWS and moved to Cloudflare too, I moved my backend domain there and that really helped with cache and eventually improved overall speed. 

ATM, I’m using two endpoints on my Nuxt3 frontend: “dynamic“ images are handled using NuxtImage and point to the Directus `/assets` endpoint, while “static“ assets use the public R2 link instead

**Reply by martinsotirov** [reply]
@luigigorlero thanks for the reply. I think I'll try to put bunny.net CDN in front of my public R2 bucket just so I can resize and optimize my images. I was hoping to be able to use the transformations from directus but since it's not possible, this seems like the next best thing.

**Comment by kolvin** [comment]
Yeah I hit this problem too, I'm self-hosting on my RPI cluster but I don't want to handle asset requests as that won't scale image transformations

Two workarounds I've considered, both of which I'm not keen on
1. Source the CDN URL directly into my content
2. Build some function to parse content at build time (currently using SSG+astro)

There was a previous discussion on this topic -> https://github.com/directus/directus/discussions/7612  but unfortunately, the proposed MR https://github.com/directus/directus/pull/8065 was closed due to insufficient capacity

The team has been reviewing their funding and licensing options -> https://github.com/directus/directus/discussions/17977 which is fair dos

Ethier need to wait on capacity or sponsor the request https://github.com/directus/directus/discussions/7612#discussioncomment-4403865

**Comment by rijkvanzanten** [comment]
@Tim-Ganther  / @kolvin, what would be your "ideal" workflow for dealing with thumbnails? The `/assets` endpoint is currently handling permissions checks and the generation of the thumbnail when it doesn't exist. The way I use it in production is by having a CloudFront distribution in front of Directus itself, so you still get the asset performance you're looking for, while keeping the ability to have private files and auto-generated thumbnail sizes. In the OP, it kinda sounds like the idea was to have the assets endpoint auto-redirect to a public URL if the file exists? Just trying to wrap my head around the ideal workflow here :)

**Reply by kolvin** [reply]
Hey @rijkvanzanten thanks for the swift response!

For context, im self hosting an instance to power a personal blog, so far, my flow is
1. upload some content to the directus media libary, storing in `s3+cloudfront`, `cloudinary` etc
2. create a new `post` entity
3. Using the markdown editor, reference an image, which embeeds a URl using the `/assets/` endpoint
   a. ![Screenshot 2023-04-27 at 17 58 13](https://user-images.githubusercontent.com/15124052/234935632-43f4958a-be84-49dd-b435-d98bad6113f9.png)

Trouble is, I want to avoid having my self hosted instance handle image requests/transformations as I'm limited by compute resource and im concered about scaling.

So for me, this currently means I have to avoid using the `/assets` endpoint + Markdown image import UI, for displaying images in blog content, as those requests will rightly, be routed to my self hosted instances instead of the public CDN URL. This is where I related to https://github.com/directus/directus/discussions/7612

Here is a really quick diagram I threw togther that hopfully helps convey what im aiming for
![self-hosted-directus(1)](https://user-images.githubusercontent.com/15124052/234954708-764e0d28-1d4d-4f6c-bef5-55e9e4c3bbf5.jpg)

Basically i think being able to define and access an asset's public URL, from a CDN, within the image import UI, would greatly help

Otherwise, I'm manually setting this or writing a client-side parsing function 🤔

**Reply by rijkvanzanten** [reply]
Hmmm yeah, the problem with that approach though is that Directus than has to assume all images are public, as you're completely bypassing the permissions system. It also means that we can't really know whether or not a thumbnail size even exists, as they're currently generated on the fly 🤔 

Maybe there's an in-between, where the /assets endpoint returns an early 301 redirect if the file exists to the cdn?

**Reply by luigigorlero** [reply]
> Maybe there's an in-between, where the /assets endpoint returns an early 301 redirect if the file exists to the cdn?

Hello! Well, that is exactly what I had in mind but I was not sure it made much sense performance-wise. Do you think it could be a walkable path? The original reason behind the post was to get insights on other production setups using Cloudfront, but a “native” Directus solution would be sparkles!

**Reply by rijkvanzanten** [reply]
> Hello! Well, that is exactly what I had in mind but I was not sure it made much sense performance-wise. Do you think it could be a walkable path?

I honestly don't know! If you go that level of complexity just to use a CDN for files, you might as well just put a CDN in front of Directus 🤔

**Comment by jcam** [comment]
Somewhat related, I also use S3 + Cloudfront for serving images. Permission controls to the images are currently open/public, but will eventually depend on a valid request token, generated on the NextJS application side. So, the main thing that I need updated in Directus is to be able to override the asset url to a custom value.
Our Directus instance itself is not publicly available, and any page updates are handled via NextJS pre-rendering.
Right now we have custom code that forcibly replaces any reference to an https://directus.example.com/assets/a12345 url and replaces it with our https://assets-cdn.example.com/project/a12345.jpg CDN endpoints, which is a bit messy and fragile.
For thumbnail generation, I have a fork of Image-Flex deployed to our Cloudfront. It generates images on the fly in the same way as Directus (https://github.com/jcam/Image-Flex).

So for me, the simplest way forward with S3+Cloudfront+self-hosted is to use the above Image-Flex asset generation tools and give the ability to override the default asset URL for Directus.

**Reply by rijkvanzanten** [reply]
So in your particular use case, you effectively want to be able to configure how URLs are saved inside a WYSIWYG editor?

**Reply by jcam** [reply]
Yes, sounds right. Our editorial team wants to be able to use the Directus built-in WYSIWYG editors to import images and drop them into the pages, Directus puts the files on S3 automatically with the S3 storage plugin, but right now <directus_url>/assets is hardcoded into the editor.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by steffenstolze** [comment]
@luigigorlero 

I'm facing the same scenario. Why don't you just put Directus in front of a CloudFront distribution?
I just tried this and it works fine. I've added a dedicated behaviour for the `/assets/*` path with a relatively large minimum TTL. This way, whenever my frontend requests a (transformed) image from Directus, it goes through CloudFront. In case of a cache miss, Directus makes the response and CloudFront caches it. When the next requests (with the same query string of course) hits the API, it's been served by CloudFront's cache (if within cache TTL).

How did you solve it for your use-case?

Cheers!

**Reply by rijkvanzanten** [reply]
This is exactly how we set it up and how I'd recommend orchestrating Directus 👍🏻
